espnet3/runner/base_runner.py

```diff
@@ -0,0 +1,392 @@
+from uuid import uuid4
+
+from dask.utils import tmpfile
+from omegaconf import OmegaConf
+from tqdm import tqdm
+
+from espnet3.parallel.parallel import (
+    get_client,
+    get_parallel_config,
+    make_client,
+    parallel_for,
+)
+from espnet3.runner.env_provider import EnvironmentProvider
+
+
+@dataclass(frozen=True)
+class AsyncJobSpec:
+    """Specification of an asynchronous shard submitted to the Dask cluster.
+
+    Attributes:
+        runner_cls (str): Import path to the concrete ``BaseRunner`` subclass.
+        provider_cls (str): Import path to the concrete ``EnvironmentProvider`` subclass
+        config (Dict): A resolved, plain-``dict`` serialization of the Hydra config.
+        params (Dict): Extra parameters forwarded to the provider/environment.
+        indices (List[int]): The subset of indices this shard should process.
+        world_size (int): Total number of shards.
+        world_rank (int): The shard rank (0..world_size-1).
+        result_path (str | None): If set, results are written to this JSONL file
+            *on the worker* and not returned to the driver.
+        extras (Dict | None): Reserved for future extensions.
+
+    Notes:
+        - Each shard is independent and reconstructs both provider and runner
+          by import path on the worker before executing :py:meth:`BaseRunner.forward`.
+    """
+
+    runner_cls: str  # e.g., module.path.to.your.RunnerClass
+    provider_cls: str  # e.g., module.path.to.your.ProviderClass
+    config: Dict
+    params: Dict
+    indices: List[int]
+    world_size: int
+    world_rank: int
+    result_path: str | None
+    extras: Dict | None
+
+
+def _import_obj(class_path: str):
+    mod, _, name = class_path.rpartition(".")
+    return getattr(importlib.import_module(mod), name)
+
+
+def _default_chunk(indices: Sequence[int], num_chunks: int) -> List[List[int]]:
+    n = max(1, num_chunks)
+    L = len(indices)
+    q, r = divmod(L, n)
+    out = []
+    start = 0
+    for i in range(n):
+        size = q + (1 if i < r else 0)
+        out.append(list(indices[start : start + size]))
+        start += size
+    return [c for c in out if c]
+
+
+def convert_paths(obj):
+    if isinstance(obj, dict):
+        return {k: convert_paths(v) for k, v in obj.items()}
+    elif isinstance(obj, list):
+        return [convert_paths(v) for v in obj]
+    elif isinstance(obj, Path):
+        return str(obj)
+    else:
+        return obj
+
+
+def get_job_cls(cluster, spec_path=None):
+    parent_cls = cluster.job_cls
+    assert spec_path is not None
+
+    class ASyncRunnerJob(parent_cls):
+        def __init__(self, *args, worker_extra_args=None, **kwargs):
+            self._user_worker_extra_args = worker_extra_args or []
+            super().__init__(*args, worker_extra_args=worker_extra_args, **kwargs)
+            python = sys.executable
+            current_file_path = Path(os.path.abspath(__file__)).resolve()
```

Comment:
`Path.resovle` will make the path absolute.

```suggestion
            current_file_path = Path(__file__).resolve()
```

Reply:
Thank you!! I will fix this